docs: adds Multi-Cluster Kubernetes Explained  pillar page (#988)
cilium/cilium.io: src/posts/2026-06-13-multi-cluster-kubernetes-explained/index.md

---
path: /blog/2026/06/13/multi-cluster-kubernetes-explained
date: 2026-06-13T14:00:00.000Z
title: 'Multi-Cluster Kubernetes Explained'
isFeatured: false
ogImage: images/multi-cluster-cover-image.jpg
ogSummary: 'Discover the core architectures and strategies behind Multi-Cluster Kubernetes. This guide covers cluster federation, cross-cluster service meshes, high availability, and how to scale and secure traffic across multiple cloud regions.'

categories:

>   - Architecture

Reviewer: ```suggestion
```

- **Hard Multitenancy**: While namespaces provide soft isolation, separate clusters offer a stronger security boundary for different teams, environments (Dev/UAT/Prod), or sensitive workloads.
- **Organizational Growth and Acquisitions**: As companies scale organically or engage in Mergers and Acquisitions (M\&A), they frequently inherit disparate infrastructure, different cloud providers, or existing Kubernetes environments. Instead of undertaking a massive, high-risk migration to consolidate these discrete workloads into a single monolith, connecting them via a multi-cluster architecture provides a seamless, non-disruptive path to infrastructure integration while allowing acquired teams to maintain their operational autonomy.

# **II. Multi-Cluster Architecture Patterns**

Organizations deploy Kubernetes across multiple clusters for various reasons such as resilience, compliance, regulatory isolation, team autonomy, or geographic distribution. Before choosing a topology, it's important to understand the trade-offs each pattern introduces in terms of operational complexity, blast radius, latency, and cost.

## **Active-Active**

> In an active-active topology, two or more clusters simultaneously serve live production traffic. Workloads are replicated across all clusters, and load is balanced between them. This pattern reduces blast radius and enables failover, supporting resilience through active-active approaches. With Cilium ClusterMesh, global services can span clusters so that pods in Cluster A and Cluster B both act as backends for the same service traffic flows to whichever is healthy.

Reviewer: ```suggestion
In an active-active topology, two or more clusters simultaneously serve live production traffic. Workloads are replicated across all clusters, and load is balanced between them. This pattern reduces blast radius and enables failover, supporting resilience through active-active approaches. With Cilium ClusterMesh, global services can span clusters so that pods in Cluster A and Cluster B both act as backends for the same service; traffic flows to whichever is healthy.
```

2. **Node-to-node IP Connectivity**: The worker nodes in all participating clusters must be able to reach each other over the network. Clusters are connected and routed across cloud providers using native VPC peering, dedicated interconnects (like AWS Direct Connect or Azure ExpressRoute), or via secure IPSec/WireGuard tunnels.
3. **Native Routing CIDR**: When using direct routing instead of encapsulation (tunneling), the network infrastructure must know how to route the Pod CIDRs. This requires configuring the underlying VPC network to recognize the pod IP ranges so that traffic can flow natively between nodes across different clusters.
4. **Consistent Datapath Mode:** While it is possible to mix some configurations, it is highly recommended that meshed clusters use a consistent datapath mode (e.g., all clusters using VXLAN/Geneve tunneling, or all clusters using direct native routing). This reduces troubleshooting complexity and ensures consistent performance.

> 5. **DNS Across Clusters:** For seamless service discovery, the Kubernetes DNS service (like CoreDNS) needs to be aware of the ClusterMesh. Cilium integrates with Kubernetes DNS to ensure that a request for a global service correctly resolves to available backend pods, regardless of which cluster they reside in.

Reviewer: ```suggestion
5. **DNS Across Clusters:** For seamless service discovery, the Kubernetes DNS service (like CoreDNS) needs to be aware of the ClusterMesh. Cilium integrates with the service routing plane underneath Kubernetes DNS. When CoreDNS resolves a global service hostname to a local virtual ClusterIP, Cilium's eBPF layer translates that IP target into a list of globally available backend pods, regardless of which cluster they reside in.
```
